doc/designs/quic-design/quic-api.md

```diff
@@ -0,0 +1,617 @@
+
+**TBD:** How should `SSL_clear` be implemented? Either:
+
+  - Modernised implementation which resets everything, handshake layer
+    re-instantiated (safer);
+  - Preserve `SSL_clear` semantics at the handshake layer, reset all QUIC state
+    (`QUIC_CHANNEL` torn down, CSM reset).
+
+#### `SSL_set0_rbio`, `SSL_set0_wbio`, `SSL_set_bio`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Changed   | Never         | No        | C             |
+
+Sets network-side BIO.
+
+The changes to the semantics of these calls are as follows:
+
+  - The BIO MUST be a BIO with datagram semantics.
+
+  - If the BIO is non-pollable (see below), application-level blocking mode will
+    be forced off.
+
+#### `SSL_set_[rw]fd`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Changed   | Never         | No        | C             |
+
+Sets network-side socket FD.
+
+Existing behaviour: Instantiates a `BIO_s_socket`, sets an FD on it, and sets it
+as the BIO.
+
+New proposed behaviour:
+
+- Instantiate a `BIO_s_dgram` instead for a QUIC connection SSL object.
+- Fails (no-op) for a QUIC stream SSL object.
+
+#### `SSL_get_[rw]fd`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Never         | No        | C             |
+
+Should not require any changes.
+
+#### `SSL_CTRL_MODE`, `SSL_CTRL_CLEAR_MODE`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| Unchanged | Never         | No        | CS            |
+
+#### SSL Modes
+
+- `SSL_MODE_ENABLE_PARTIAL_WRITE`: Implemented. If this mode is set during a
+  non-partial-write `SSL_write` operation spanning multiple `SSL_write` calls,
+  this operation is aborted and partial write mode begins immediately.
+
+- `SSL_MODE_ACCEPT_MOVING_WRITE_BUFFER`: Implemented.
+
+- `SSL_MODE_AUTO_RETRY`: TBD.
+
+- `SSL_MODE_RELEASE_BUFFERS`: Ignored. This is an optimization and if it has
+  any sensible semantic correspondence to QUIC, this can be considered later.
+
+- `SSL_MODE_SEND_FALLBACK_SCSV`: TBD: Either ignore or fail if the client
+  attempts to set this prior to handshake. The latter is probably safer.
+
+  TBD: What if the client attempts to set this post handshake? Ignore it?
+
+- `SSL_MODE_ASYNC`: TBD.
+
+### New APIs
+
+TBD: Should any of these be implemented as ctrls rather than actual functions?
+
+#### `SSL_tick`
+
+| Semantics | `SSL_get_error` | Can Tick? | CSHL          |
+| --------- | ------------- | --------- | ------------- |
+| New       | Never         | Yes       | CS            |
+
+Advances the QUIC state machine to the extent feasible, potentially performing
+network I/O. Also compatible with DTLSv1 and supercedes `DTLSv1_handle_timeout`
+for all use cases.
```

Comment:
SSL_tick() is needed to handle time based events. It is not sufficient just to rely on SSL_do_handshake(), SSL_read() and SSL_write() internally calling tick. Periodically we will need to send re-transmissions etc after some given timeout

Reply:
> And so I think the current SSL_tick() documentation may need an update, to the point I currently fail to see why we have SSL_tick() function other than that DTLSv1_handle_timeout() is a bad name for QUIC.

If a client application frequently calls SSL_read and SSL_write there is no issue because libssl is getting plenty of opportunities to handle any upcoming timeout events and process incoming data, etc. But it is entirely valid usage for a client to open a connection and then not call SSL_read or SSL_write at all for a long time. In this use case it is important that SSL_tick be called to give us a chance to process new incoming packets, handle timeout events, etc.

Note that these functions replace equivalent calls in our DTLSv1 code, so there is basically strong prior art for the need here.

Having said that, as far as documentation is concerned "the customer is always right" &mdash; if the purpose of SSL_tick is not clear from reading the current documentation clearly I need to improve it, so I'll make this a TODO.